notation.go

```diff
@@ -65,6 +68,12 @@ type Signer interface {
 	Sign(ctx context.Context, desc ocispec.Descriptor, opts SignerSignOptions) ([]byte, *signature.SignerInfo, error)
 }
 
+type BlobSigner interface {
```

Comment:
I'm thinking if we could reuse the existing `Signer` interface, `SignOptions` struct and func `notation.Sign`, so the design could be:
```
// SignOptions contains parameters for notation.Sign.
type SignOptions struct {
	SignerSignOptions

        // BlobSign, if set to true, specifies a blob signing
        BlobSign bool

        // BlobSignDescriptor is the blob descriptor to be signed.
        // It is only used when 'BlobSign' is true.
        BlobSignDescriptor *ocispec.Descriptor

	// ArtifactReference sets the reference of the artifact that needs to be
	// signed. It can be a tag, a digest or a full reference.
        // It is only used when 'BlobSign' is false.
	ArtifactReference string

	// UserMetadata contains key-value pairs that are added to the signature
	// payload
	UserMetadata map[string]string
}
```
In the above design, we are adding new fields into existing struct without any breaking change. Callers just need to set the flag to choose between signing a blob and an oci image. 
Callers are also allowed to provide their own descriptor to get signed. For example, Notation CLI will create such a descriptor out of the `blob path` and pass it to notation-go.

Reply:
I think we can drop the `BlobSigner` interface and reuse the `Signer` interface as it is generic enough to cover all signing scenarios. Meanwhile, I think we can keep the `notation.SignBlob()` method and `notation.SignBlobOptions` type as utilities.

Reply:
Since the [blob sign spec](https://github.com/notaryproject/notation/pull/811) is still under review, it would be better revisit this after the spec is merged.

Reply:
The problem with using [Signer Interface](https://github.com/notaryproject/notation-go/blob/main/notation.go#L62-L66) because the `Sign(ctx context.Context, desc ocispec.Descriptor, opts SignerSignOptions) ([]byte, *signature.SignerInfo, error)` method takes descriptor as argument but in case of blob we would need to create descriptor from `byte[]`. To create descriptor we need to know the digest algo and for that we need to call plugin for [`DescribeKey`](https://github.com/notaryproject/notation-go/blob/main/signer/plugin.go#L94-L97) and since we dont have access to plugin in `notation.go` we cannot use SignerInterface.

Reply:
@priteshbandi Just sharing my thoughts here:
1. User specifies the <blob_path> + <hash_alg> to Notation CLI, where <hash_alg> is used to hash the blob content, supported values are `SHA256`, `SHA384` and `SHA512`.
2. Notation CLI constructs the blob descriptor based on user inputs. An example desc would be:
```
{
    "mediaType": "application/octet-stream",
    "digest": digest.FromBytes(hash(blob_content)),
    "size": 1024,
    "annotations": { // user defined metadata
        "hash-algorithm": SHA512
    }
}
```
Since the `hash-algorithm` is also in the payload, it's get signed as well and integrity is guaranteed.
4. Notation CLI uses the same `Signer interface` and call `notation-go SignBlob` method to sign the descriptor.
5. On verification, the <hash-algorithm> in the payload is used to hash the content then compute the digest again and compare with the value in the target descriptor.

(Similar design was seen in a Timestamp countersignature, RFC 3161.)

Reply:
As we discussed offline, just re-posting my two cents here regarding the hash algorithm:
I think we don't have to use the `signing algorithm` to hash the blob content when computing the `digest` field of the target descriptor. This wasn't the case when we sign an `oci image` either. When we sign an `oci image`, we resolve the descriptor from the reference without knowing any information of the signing key. Then we send the descriptor along with other signed attributes to the `Signer`, who uses the `signing algorithm` to hash the payload again before signing.
Similarly, for `blob signing` we could use [digest.FromReader](https://pkg.go.dev/github.com/opencontainers/go-digest@v1.0.0#FromReader) with [Canonical](https://pkg.go.dev/github.com/opencontainers/go-digest@v1.0.0#Canonical) hash algorithm, essentially SHA256. This user experience aligns with what we have today. 

In terms of hash collision, SHA256 is a collision-resistant hash. Also, there is the `size` field, which is hashed by the `signing algorithm` as a part of the final payload before generating the signature. Therefore, it is even harder to find such a collision under SHA256 AND with the same size.

Reply:
I think we don't have to use the signing algorithm to hash the blob content when computing the digest field of the target descriptor. This wasn't the case when we sign an oci image either. When we sign an oci image, we resolve the descriptor from the reference without knowing any information of the signing key. Then we send the descriptor along with other signed attributes to the Signer, who uses the signing algorithm to hash the payload again before signing.

> In my opinion, we shouldn't draw a comparison between image signing and blob signing. In the case of OCI images, the descriptor itself is the content being signed (registry decides if they want to represent an artifact using sha256 or sha512 hash). However, with blobs, we first need to create the descriptor, which is then used for signing.  


Similarly, for blob signing we could use [digest.FromReader](https://pkg.go.dev/github.com/opencontainers/go-digest@v1.0.0#FromReader) with [Canonical](https://pkg.go.dev/github.com/opencontainers/go-digest@v1.0.0#Canonical) hash algorithm, essentially SHA256. This user experience aligns with what we have today. The design becomes pretty simple with much lower engineering cost. (I'm seeing duplicate codes in the implementation of the new BlobSigner interface as the current Signer interface is already very generic.)

> Yes but as per [signature-specification.md#algorithm-selection](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#algorithm-selection), it is recommended to use a larger hashing algorithm for higher key sizes. In my opinion, we should adhere to this practice as it prevents confusion (as its similar to Image signing), and users with larger keys would naturally anticipate enhanced signature strength. Additionally, the process of hashing with one algorithm for descriptor generation and then employing another algorithm for signature generation seems counterintuitive.

Reply:
@priteshbandi Just want to clarify, the your proposed signing process actually hashes 3 times, they are:
1. Using the signature algorithm associated hash func to compute the hash of the blob content. https://github.com/notaryproject/notation-go/blob/1263e20b887f51ca7e37eb2009fe54c798385c09/signer/plugin.go#L253
2. Using `digest.FromBytes(<hash computed from step 1>)`, this step uses the [Canonical hash](https://github.com/opencontainers/go-digest/blob/ea51bea511f75cfa3ef6098cc253c5c3609b037a/digest.go#L93) to compute the digest field of the target descriptor. https://github.com/notaryproject/notation-go/blob/1263e20b887f51ca7e37eb2009fe54c798385c09/signer/plugin.go#L267
3. The final step before signature generation, the `Signer` uses the signing algorithm to hash the payload again.

My suggestion is removing step 1 and change step 2 to just `digest.FromReader`.

Reply:
> @priteshbandi Just want to clarify, your proposed signing process actually hashes 3 times, they are:
Nop, I am advocating about using the same hashing algorithm throughout the signing operation. In this case it would be 1) during content digest calculation to generate descriptor 2)  during the actual signature generation process, the Signer uses the signing algorithm to hash the payload again.
> I dont think we need 1st step.

>This is quite common actually, see RFC 5652 CMS as an example: https://www.rfc-editor.org/rfc/rfc5652#section-5.3. Under Signed-data type, there's the digestAlgorithm and signatureAlgorithm defined separately.
Indeed, the specification permits all these permutations, but I don't see a compelling reason  to not maintain consistency in the signing algorithm throughout the entire signing operation. Having a consistent hashing algo will not only improve user experience as they wont need to worry about passing additional attribute (or introducing another defualt) but will also keep it consistent with container image signing spec. While I acknowledge that this may introduce some code complexity, in my opinion, it's a worthwhile tradeoff

Reply:
> Nop, I am advocating about using the same hashing algorithm throughout the signing operation. In this case it would be 1) during content digest calculation to generate descriptor 2)  during the actual signature generation process, the Signer uses the signing algorithm to hash the payload again.
> I dont think we need 1st step.

But your current implementation in this PR does not align with what you are stating above:  
the current implementation in this PR does hash 3 times. `digest.FromBytes` in the `getDescriptor` function, which uses SHA256 as hash func.

Reply:
But your current implementation in this PR does not align with what you are stating above: the current implementation in this PR does hash 3 times. `digest.FromBytes` in the `getDescriptor` function, which uses SHA256 as hash func.

Ohh, that's a bug in code :D; Also, this PR was raised to build consensus on implementation approach and is not ready for detailed review.

Reply:
Got it. Please fix the bug. Ideally, we should only need to hash 2 times instead of 3.

Reply:
Done.

Reply:
> Since the [blob sign spec](https://github.com/notaryproject/specifications/pull/283) is still under review, it would be better revisit this after the spec is merged.

By following my above comment as well as @sajayantony 's comment in [the community meeting (2/5/2024)](https://hackmd.io/_vrqBGAOSUC_VWvFzWruZw?both#Feb-5-2024), let's follow up this conversion in the spec PR.